src/charm.py

```diff
@@ -43,7 +46,16 @@ def __init__(self, *args: Any) -> None:
         except CharmConfigInvalidError as exc:
             self.model.unit.status = ops.BlockedStatus(exc.msg)
             return
-        self._synapse = Synapse(charm_state=self._charm_state)
+        self._database = DatabaseObserver(self)
+        self.framework.observe(
+            self._database.database.on.database_created, self._on_database_created
+        )
+        self.framework.observe(
```

Comment:
Any thoughts on the observer pattern for the database?

Reply:
It was my first try but then I couldn't solve how to call the config-changed method from the Observer class. That's why I moved back to the charm.py. Any ideas on how could I do that from the Observer class?

Reply:
@yanksyoon?

Reply:
Considering the discussion [here](https://github.com/canonical/flask-k8s-operator/pull/14#discussion_r1203524175) in Discourse and the example in [Flask](https://github.com/canonical/flask-k8s-operator/blob/main/src/databases.py#L59), the solution would replan the charm in the Observer but in this case, there is a call to migrate_config (a command to regenerate the command) so I'm afraid that I would need to replicate the config-change code on the Observer.

IMHO I don't see much gain in doing that. Couldn't we keep the observer in the charm and apply the Observer pattern in cases where there is no need for layer replan?

Reply:
@gregory-schiano probably one for you to take a look at

Reply:
And what would the circular import be please?

Reply:
I thought that would be in my previous commits but it wasn't so I created this branch to show what I tried:
https://github.com/canonical/synapse-operator/blob/postgresql-observer-pattern/src/database.py#L62

Reply:
@yanksyoon Thanks for the suggestion. I applied the idea of PebbleService here then I could remove the observer from the charm.